libs/langgraph/langgraph/prebuilt/tool_node.py

```diff
@@ -67,13 +69,75 @@ def msg_content_output(output: Any) -> str | List[dict]:
             return str(output)
 
 
+def _handle_tool_error(
+    e: Exception,
+    *,
+    flag: Union[
+        bool,
+        str,
+        Callable[..., str],
+        tuple[type[Exception], ...],
+    ],
+) -> str:
+    if isinstance(flag, (bool, tuple)):
+        content = TOOL_CALL_ERROR_TEMPLATE.format(error=repr(e))
+    elif isinstance(flag, str):
+        content = flag
+    elif callable(flag):
+        content = flag(e)
+    else:
+        raise ValueError(
+            f"Got unexpected type of `handle_tool_error`. Expected bool, str "
+            f"or callable. Received: {flag}"
+        )
+    return content
+
+
+def _infer_handled_types(handler: Callable[..., str]) -> tuple[type[Exception]]:
+    sig = inspect.signature(handler)
+    params = list(sig.parameters.values())
+    if params:
+        # If it's a method, the first argument is typically 'self' or 'cls'
+        if params[0].name in ["self", "cls"] and len(params) == 2:
+            first_param = params[1]
+        else:
+            first_param = params[0]
+
+        type_hints = get_type_hints(handler)
+        if first_param.name in type_hints:
+            if get_origin(first_param.annotation) is Union:
+                return tuple(get_args(first_param.annotation))
+            return (type_hints[first_param.name],)
+
+    # If no type information is available, return (Exception,) for backwards
+    # compatibility.
+    return (Exception,)
+
+
 class ToolNode(RunnableCallable):
     """A node that runs the tools called in the last AIMessage.
 
     It can be used either in StateGraph with a "messages" key (or a custom key passed via ToolNode's 'messages_key').
     If multiple tool calls are requested, they will be run in parallel. The output will be
     a list of ToolMessages, one for each tool call.
 
+    Args:
+        tools: A sequence of tools that can be invoked by the ToolNode.
+        name: The name of the ToolNode in the graph. Defaults to "tools".
+        tags: Optional tags to associate with the node. Defaults to None.
+        handle_tool_errors: How to handle tool errors raised by tools inside the node. Defaults to True.
```

Comment:
Is `True` the safe default? I know it's convenient, but what's the contents of `error` in the error message?

Is it the full stack trace? Just the error message? Could something sensitive be leaked by mistake as part of the error message? (e.g., API Key)

Reply:
it's the current default, so i think it would be problematic to change it as it would be breaking. the content is just the error message, no stack trace

hm, that's an interesting point about a sensitive info in error message -- possible? but IMO this is the same class of issue as "what if someone invokes a tool/graph with their API key as input"

Reply:
if it's default can leave as is